### PR TITLE
improve compatibly with node.js ESM

### PR DESCRIPTION
**Pull request by sokra**
improve experiments.mjs to be more compatibly with node.js ESM
  - add support for `type: "module"` in package.json
  - add support for `*.cjs`
  - enable strict ESM for data uris with `text/javascript` or `application/javascript`
  - disallow not fully specified requests in imports `*.mjs` or `type: "module"`

add `descriptionData` rule condition to match package.json info
fix passing resolveOptions along context modules

**What kind of change does this PR introduce?**
feature, experiment

**Did you add tests for your changes?**
yes

**Does this PR introduce a breaking change?**
yes to the experiment

**What needs to be documented once your changes are merged?**
Using `*.js` with `"type": "module"` in package.json or `*.mjs` moves webpack into a strict ESM mode where more limitations are applied:
* Importing named exports from non-esm modules (e. g. CommonJs, JSON) is no longer possible. Only the default export can be used.
  * `import * as X from "commonjs"` creates a namespace object with default export instead of behaving like `const X = require("commonjs")
* Requests has to be fully specified, with filename and extension (expect for bare modules without subpath)
* CommonJs Syntax is not available: `require`, `module`, `exports`, `__filename`, `__dirname`

### CODE REVIEW

File: lib/rules/RuleSetCompiler.js [hunk 1]
```diff
@@ -76,10 +76,26 @@ class RuleSetCompiler {
 		const execRule = (data, rule, effects) => {
 			for (const condition of rule.conditions) {
 				const p = condition.property;
-				if (p in data) {
+				if (Array.isArray(p)) {
+					let current = data;
+					for (const subProperty of p) {
+						if (typeof current === "object" && subProperty in current) {
```
Comment (vankop) [inline lib/rules/RuleSetCompiler.js:82]: check `null`?

File: lib/config/defaults.js [hunk 2]
```diff
@@ -364,16 +367,56 @@ const applyModuleDefaults = (
...
+				{
+					test: /\.mjs$/i,
+					...esm
+				},
+				{
+					test: /\.js$/i,
+					descriptionData: {
+						type: "module"
+					},
+					...esm
```
Comment (vankop) [inline lib/config/defaults.js:396]: So to sum up, to enable "javascript/esm" I need add `experiments: {mjs: true}` and `type: module` in description file?

Comment (vankop) [reply lib/config/defaults.js:396]: maybe add another experiment for this?

Comment (sokra) [reply lib/config/defaults.js:396]: `*.mjs` and `type: "module"` is both part of the node.js behavior.

I actually think about removing the experiment and enable it by default. I'm a bit worries about node.js ESM support is still flagged as experimental...

**Pull request merged**